tools/closed-by-commits

```diff
@@ -0,0 +1,434 @@
+
+        repository = self.github.get_repo(f"zulip/{self.reponame}")
+        comparison = repository.compare(base, head)
+        commit_shas = [commit.sha for commit in comparison.commits]
+        console.log(f"Found {len(commit_shas)} commits")
+
+        all_seen_prs = set()
+        issue_to_prs: dict[Issue, set[PullRequest]] = defaultdict(set)
+        with Progress(console=console) as progress:
+            task = progress.add_task("Processing commits...", total=len(commit_shas))
+            for sha in commit_shas:
+                progress.console.log(f"Processing {sha}")
+
+                this_issue_to_prs, this_seen_prs = self.get_issues_for_commit(sha)
+                for issue, prs in this_issue_to_prs.items():
+                    issue_to_prs[issue].update(prs)
+                all_seen_prs.update(this_seen_prs)
+                progress.advance(task)
+
+        for issue, prs in issue_to_prs.items():
+            for pr in prs:
+                issue.czo_urls.update(pr.czo_urls)
+
+            issue.closed_by_prs = sorted(prs, key=lambda p: p.number)
+
+        unique_prs = len({pr for prs in issue_to_prs.values() for pr in prs})
+        console.log(
+            f"Found {len(all_seen_prs)} PRs, with {unique_prs} unique PRs closing {len(issue_to_prs)} issues",
+        )
+
+        return sorted(issue_to_prs.keys(), key=lambda x: x.number)
+
+
+def validate_github_token(value: str) -> str:
+    if not value.startswith("github_"):
+        raise typer.BadParameter("Github access tokens start with `github_`")
+    return value
+
+
+from enum import Enum
+
+
+class ZulipRepo(str, Enum):
+    zulip = ("zulip",)
+    flutter = "zulip-flutter"
+
+
+repo_issue_prefixes = {
+    ZulipRepo.zulip: "",
+    ZulipRepo.flutter: "F",
+}
+
+
+def main(
+    base_commit: Annotated[str, typer.Argument(help="Commit-ish, resolved on the server")],
+    head_commit: Annotated[str, typer.Argument(help="Commit-ish, resolved on the server")],
+    token: Annotated[
+        str,
+        typer.Option(
+            metavar="TOKEN",
+            envvar="GITHUB_TOKEN",
+            show_envvar=True,
+            callback=validate_github_token,
+            help=(
+                "Github access token; can be a fine-grained personal access token "
+                "with read-only access to 'Public repositories'.  "
+                "See https://github.com/settings/personal-access-tokens/new"
+            ),
+        ),
+    ],
+    repo: ZulipRepo = ZulipRepo.zulip,
+) -> int:
+    """Find issues which are closed in a commit range."""
+
+    lines = []
+    prefix = repo_issue_prefixes[repo]
+    analyzer = CommitRangeAnalyzer(token, repo.value, prefix)
+    for issue in analyzer.analyze_range(base_commit, head_commit):
+        if not issue.closed_by_prs and not issue.czo_urls:
+            # Closed an issue that was never discussed anywhere?
+            continue
+
+        lines.append(f"## #{prefix}{issue.number}: {issue.title}")
```

Comment:
Bikeshed comment: The h2 feels too big to me in the output here. On a quick experiment (by editing in the compose box and previewing), I like h4 best:
```suggestion
        lines.append(f"#### #{prefix}{issue.number}: {issue.title}")
```